src/xclim/indices/_hydrology.py

```diff
@@ -647,3 +648,43 @@ def low_flow_frequency(q: xarray.DataArray, threshold_factor: float = 0.2, freq:
     threshold = threshold_factor * mean_flow
     out = threshold_count(q, "<", threshold, freq=freq)
     return to_agg_units(out, q, "count")
+
+
+@declare_units(pr="[precipitation]")
+def antecedent_precipitation_index(pr: xarray.DataArray, window: int = 7, p_exp: float = 0.935) -> xarray.DataArray:
+    """
+    Antecedent Precipitation Index.
+
+    Calculate the running weighted sum of daily precipitation values given a window and weighting exponent.
+    This index serves as an indicator for soil moisture.
+
+    Parameters
+    ----------
+    pr : xarray.DataArray
+        Daily precipitation data.
+    window : int, optional
+        Window for the days of precipitation data to be weighted and summed, default is 7.
+    p_exp : float, optional
+        Weighting exponent, default is 0.935.
+
+    Returns
+    -------
+    xarray.DataArray
+        Antecedent Precipitation Index.
+
+    References
+    ----------
+    :cite:cts:`schroter2015,li2021`
+    """
+    # TODO: ensure precip is mm
```

Comment:
This answers your first question. 

I suggest converting from a rate to an amount explicitly and then ensuring we have "mm".

```suggestion
    pr = xc.core.units.rate2amount(pr)
    pr = xc.core.units.convert_units_to(pr, 'mm', context='hydro')
```
The  `declare_units` above will make sure that `pr` was given with precipitation units, which includes two possibilities : a mass flux per area (ex: `kg m-2 s-1`) or a LWE rate (ex: `mm d-1`). With `rate2amount` we multiply the flux/rate by the length of the timestep to get an amount and then we convert to  `mm` because that's how the API defined.   Passing  `context='hydro'` will allow xclim to use the density of liquid water to convert the mass per area to a thickness.

Reply:
Thank you for explaining!